urlcaef331f4e_dreamable_cra_tgz-pJ8-coho_lp_project_LPProjectJ8/src/coho/lp/solver/DoubleIntervalCohoSolver.java

```diff
@@ -0,0 +1,521 @@
+		pos.assign(false,col);//remove that column
+		Matrix A = dualA.col(pos);
+		Matrix b = dualB;
+		Matrix c = dualC.row(pos);
+		return LP.createCohoDual(new Constraint(A,b),c);
+	}
+	
+	
+	/*
+	 * LPresult for the coho primary problem from coho dual result.
+	 */
+	private CohoSolverResult cohoResultFromDual(CohoSolverResult dualResult){
+		if(dualResult==null)
+			return null;
+		LPResult.ResultStatus status;
+		switch(dualResult.status()){
+		case OK: 
+			status=LPResult.ResultStatus.OK;
+			break;
+		case POSSIBLEOK: 
+			status=LPResult.ResultStatus.POSSIBLEOK;
+			break;
+		case INFEASIBLE:
+			status=LPResult.ResultStatus.INFEASIBLEORUNBOUNDED;
+			break;
+		case UNBOUNDED:
+			status=LPResult.ResultStatus.INFEASIBLE;
+			break;
+		case INFEASIBLEORUNBOUNDED:
+			status=LPResult.ResultStatus.INFEASIBLEORUNBOUNDED; 
+		default:
+			throw new RuntimeException("Not reachable code");	
+		}
+		if(status==LPResult.ResultStatus.OK||status==LPResult.ResultStatus.POSSIBLEOK){
+			Iterator<LPBasis> optBases = dualResult.optBasesIter();
+			ArrayList<Matrix> optPoints = new ArrayList<Matrix>();
+			while(optBases.hasNext()){
+				LPBasis basis = optBases.next();
+				try{
+					optPoints.add(cohoPoint(basis));
+				}catch(SingularMatrixException e){
+					//then don't compute the opt point. It's well-condition for dual, ill-condition for primal
+					//return new CohoSolverResult(status,dualResult.optCost().negate(),dualResult.optBases(),null);					
+					//This should never happen. Because the only place to throw SingularMatrixException is from 
+					//checkCondition of DoubleIntervalCohoMatrix. Where the condition numbers for A and A' are the same. 
+					//The cohoDual feasibility is checked before, so, there never throw an exception.
+					throw new LPError("This should never happen");
+				}
+			}
+			return new CohoSolverResult(status,dualResult.optCost().negate(),
+					dualResult.optBases(), optPoints.toArray(new Matrix[optPoints.size()]));
+		}
+		else
+			return new CohoSolverResult(status,(DoubleInterval)null,(LPBasis[])null,null);
+	}
+
+	@Override //throw exception when clear feasible basis found.
+	public CohoSolverResult cohoDualResult(LPBasis basis){
+		CohoSolverResult result = super.cohoDualResult(basis);
+		if(result!=null &&result.status()==LPResult.ResultStatus.OK){
+			throw new FindClearOptResultException(result);
+		}
+		return result;
+	}
+	// return false if it is clearly not feasible
+	// return true if it is possible feasible or clearly feasible. 
+	// This is the only place that could throw an SingularMatrixException.
+	// Because in findOpt, for each new basis, this function is called first. 
+	// If the basis is ill-condition, here will catch it and throw the IllCondBasisException
+	@Override //throw illcondition when SingularMatrixException throw from getSolution 
+	public boolean cohoDualFeasible(LPBasis basis){
+		try{
+			return super.cohoDualFeasible(basis);
+		}catch(LPError e){
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)